web/client-api/src/test/java/io/deephaven/web/client/api/AbstractAsyncGwtTestCase.java

```diff
@@ -0,0 +1,391 @@
+                }
+                complete[0] = true;
+                // complete already handled
+            }, timeout * TIMEOUT_SCALE);
+
+        });
+    }
+
+    protected <V extends HasEventHandling, T> Promise<V> waitForEvent(V evented, String eventName,
+            Consumer<CustomEvent<T>> check, int timeout) {
+        return this.<V, T>waitForEventWhere(evented, eventName, e -> {
+            check.accept(e);
+            return true;
+        }, timeout);
+    }
+
+
+    protected static <T> Promise<T> promiseAllThen(T then, IThenable<?>... promises) {
+        return Promise.all(promises).then(items -> Promise.resolve(then));
+    }
+
+    protected <T> IThenable<T> waitFor(BooleanSupplier predicate, int checkInterval, int timeout, T result) {
+        return new Promise<>((resolve, reject) -> {
+            schedule(predicate, checkInterval, () -> resolve.onInvoke(result),
+                    () -> reject.onInvoke("timeout of " + timeout + " exceeded"), timeout);
+        });
+    }
+
+    protected <T> IThenable.ThenOnFulfilledCallbackFn<T, T> waitFor(int millis) {
+        return result -> new Promise<>((resolve, reject) -> {
+            DomGlobal.setTimeout(p -> resolve.onInvoke(result), millis);
+        });
+    }
+
+    protected <T extends HasEventHandling> IThenable.ThenOnFulfilledCallbackFn<T, T> waitForEvent(T table,
+            String eventName, int millis) {
+        return result -> new Promise<>((resolve, reject) -> {
+            boolean[] success = {false};
+            table.addEventListenerOneShot(eventName, e -> {
+                success[0] = true;
+                resolve.onInvoke(table);
+            });
+            DomGlobal.setTimeout(p -> {
+                if (!success[0]) {
+                    reject.onInvoke("Waited " + millis + "ms");
+                }
+            }, millis);
+        });
+    }
+
+    private void schedule(BooleanSupplier predicate, int checkInterval, Runnable complete, Runnable fail, int timeout) {
+        if (timeout <= 0) {
+            fail.run();
+        }
+        if (predicate.getAsBoolean()) {
+            complete.run();
+        }
+        DomGlobal.setTimeout(ignore -> {
+            if (predicate.getAsBoolean()) {
+                complete.run();
+            } else {
+                schedule(predicate, checkInterval, complete, fail, (timeout * TIMEOUT_SCALE) - checkInterval);
+            }
+        }, checkInterval);
+    }
+
+    protected Promise<JsTable> assertNextViewportIs(JsTable table, Function<JsTable, Column> column,
+            String[] expected) {
+        return assertUpdateReceived(table, viewportData -> {
+            String[] actual = Js.uncheckedCast(getColumnData(viewportData, column.apply(table)));
+            assertTrue("Expected " + Arrays.toString(expected) + ", found " + Arrays.toString(actual) + " in table "
+                    + table + " at state " + table.state(), Arrays.equals(expected, actual));
+        }, 2000);
```

Comment:
Define as constant - same timeout is used in the other `assertNextViewportIs` method.

Reply:
Actually I want them to be different-ish constants, so you can nail down which timeout failed, since it can be a pain to find them.

Reply:
Yup that's fair too, though would be nice to define a base constant and just add an offset for uniqueness